contracts/exchange/contracts/src/MixinTransferSimulator.sol

```diff
@@ -0,0 +1,59 @@
+/*
+
+  Copyright 2018 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.9;
+pragma experimental ABIEncoderV2;
+
+import "./MixinAssetProxyDispatcher.sol";
+
+
+contract MixinTransferSimulator is
+    MixinAssetProxyDispatcher
+{
+    /// @dev This function may be used to simulate any amount of transfers
+    /// As they would occur through the Exchange contract. Note that this function
+    /// will always revert, even if all transfers are successful. However, it may
+    /// be used with eth_call or with a try/catch pattern in order to simulate
+    /// the results of the transfers.
```

Comment:
This is an interesting design pattern. Why is this better than reverting only when there is an error?

Reply:
I think we would need a way to detect we're in an `eth_call` context otherwise.

Reply:
Ah, that's right. Okay 👍

Reply:
Because it's a public function with access to assets, so we want to ensure that it always reverts state when called :) The purpose of this is to allow validation of arbitrary transfers within a single RPC call (people have been experiencing rate limiting issues + this will speed up validation a good amount).